Duplicati/Library/Main/Operation/Restore/BlockManager.cs

```diff
@@ -0,0 +1,400 @@
+                }
+            }
+
+            /// <summary>
+            /// Get a block from the cache. If the block is not in the cache,
+            /// it will request the block from the volume and return a `Task`
+            /// that will be completed when the block is available.
+            /// </summary>
+            /// <param name="block_request">The requested block.</param>
+            /// <returns>A `Task` holding the data block.</returns>
+            public Task<byte[]> Get(BlockRequest block_request)
+            {
+                // Check if the block is already in the cache, and return it if it is.
+                if (m_block_cache.TryGetValue(block_request.BlockID, out byte[] value))
+                {
+                    CheckCounts(block_request);
+                    return Task.FromResult(value);
+                }
+
+                // If the block is not in the cache, request it from the volume.
+                sw_get_wait?.Start();
+                var tcs = new TaskCompletionSource<byte[]>();
+                var new_tcs = m_waiters.GetOrAdd(block_request.BlockID, tcs);
+                if (tcs == new_tcs)
+                {
+                    // We are the first to request this block
+                    m_volume_request.Write(block_request);
+                }
+                sw_get_wait?.Stop();
+
+                return new_tcs.Task.ContinueWith(t => {
+                    CheckCounts(block_request);
+                    return t.Result;
+                });
+            }
+
+            /// <summary>
+            /// Set a block in the cache. If the block is already in the cache,
+            /// it will be replaced. If the block is not in the cache, it will
+            /// be added. If the block is no longer needed, it will be removed
+            /// from the cache. If the block is requested while it is being
+            /// removed, the requester will be given a `Task` that will be
+            /// completed when the block is available.
+            /// </summary>
+            /// <param name="key"></param>
+            /// <param name="value"></param>
+            public void Set(BlockRequest blockRequest, byte[] value)
+            {
+                m_block_cache.Set(blockRequest.BlockID, value);
+
+                // Notify any waiters that the block is available.
+                if (m_waiters.TryRemove(blockRequest.BlockID, out var tcs))
+                {
+                    tcs.SetResult(value);
+                }
+            }
+
+            /// <summary>
+            /// Retire the dictionary. This will decrement the number of readers
+            /// accessing the dictionary, and if there are no more readers, it
+            /// will retire the volume request channel, effectively shutting
+            /// down the restore process network.
+            /// </summary>
+            public void Retire()
+            {
+                if (Interlocked.Decrement(ref readers) <= 0) {
```

Comment:
Right, if it only had a single channel. This is due to each `FileProcessor` having it's own request/response channel set, so that each request is returned to the respective `FileProcessor`. So this count variable is used to keep track of how many readers are still active, to avoid premature shutdown of the network.